### DISCUSSION TITLE
Display Drivers and Sequences

### DISCUSSION THREAD

**Discussion by kevinsmgov** [OrchardCMS/OrchardCore#10053] [Q&A] [answered]
I'm looking for a best practice approach for this scenario:

- A display driver for a parent content item performs a query and returns a list of child content items it will include in its ViewModel (e.g. `IEnumerable<dynamic> ContentItems`)
- For each item, `BuildDisplayAsync` is called so that the child items are also shapes and can be displayed with their own template (think of a page with a series of cards, each child item called with `DisplayAsync` renders a card)
- The ViewModel for each card item needs to contain an Index property (0, 1, 2, etc.) based on their position in the collection so that different rendering can be applied based on location in the list

Since the Index property is based on the collection at runtime, it can't be part of the card's ContentType definition, only its ViewModel would contain it. The problem I'm running into is how to pass the index value to each child's `BuildDisplayAsync` call so that the child's Display Driver can include it in its own ViewModel.

Any suggestions on how best to approach this?

**Comment by deanmarcussen** [comment, marked as answer]
Set the property on the view model after the call to build display.

there’s no way to pass arguments to a build display call

(you could do also putting the index in a httpcontext value , and retrieving it, but I would do the first suggestion)

the reason is because build display should not depend on external parameters, other than the given model

**Reply by kevinsmgov** [reply]
I had that in place before I added a display driver for the child item (it was just using the default content item driver).

I’m trying to move towards strongly typed templates so I created a display driver. Now the logic to add the Index property after the call to `BuildDisplayAsync` doesn’t work anymore.

Are you aware of any examples of adding data to a strongly typed shape after the call to `BuildDisplayAsync`?

**Reply by deanmarcussen** [reply]
Without seeing any code I would assume your strongly typed display driver is still placing its shape in a zone, probably a Content Zone so what you might be experiencing is that the property is being added to the zone rather that the specific shape your display driver is rendering.

In which case find the shape you want to add the property to, by looking at the zones

**Reply by kevinsmgov** [reply]
I spent a couple hours this morning debugging through things before I posted this discussion and somehow missed the "Items" collection on shape.Content where my returned Castle.Proxies._ViewModelProxy_ lived. That's what I was looking for. Thanks for your time.

**Reply by kevinsmgov** [reply]
For those interested, here's a code snippet from a display driver that returns a collection of strongly typed shapes and then modifies them before adding them to the resulting view model. In this case, an Index property contains the ordinal value of the child item in the parent collection so it's available to its template.
```
var contentItems = await _contentManager.GetAsync(part.PromotionsList.ContentItemIds);
var shapes = await Task.WhenAll(contentItems.Take(max).Select(async item => await _contentItemDisplayManager.BuildDisplayAsync(item, context.Updater, displayType)));
    for (var index = 0; index < shapes.Length; index++)
    {
        var shape = shapes[index] as dynamic;
        var items = shape.Content.Items as IReadOnlyList<IPositioned>; // one or more Castle.Proxies specific to the view model names
        var proxy = items.SingleOrDefault(item => item.GetType().FullName == "Castle.Proxies.PromotionViewModelProxy") as dynamic;
        if(proxy != null)
            proxy.Index = index; // the PromotionViewModel for the child items has an Index property for the ordinal value
    }
viewModel.ContentItems = shapes;
```
I believe the Castle.Proxies classes are dynamically created so I don't think there's a way to make the last two lines strongly typed. Let me know if you find out otherwise.
